### PR TITLE
Add column axis in ks.concat

### LINKED ISSUE #625: concat(axis=1) currently is not supported
Concatenating along the columns is a common use case but currently isn't supported in koalas. We can achieve this by joining the dataframes by their index.

```
>>> df4 = pd.DataFrame([['bird', 'polly'], ['monkey', 'george']],
...                    columns=['animal', 'name'])
>>> pd.concat([df1, df4], axis=1)
  letter  number  animal    name
0      a       1    bird   polly
1      b       2  monkey  george
```

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR proposes to add `axis` at `ks.concat(...)`.

```python
import databricks.koalas as ks

df1 = ks.DataFrame([['a', 1], ['b', 2]], columns=['letter', 'number'])
df2 = ks.DataFrame([['bird', 'polly'], ['monkey', 'george']], columns=['animal', 'name'])
ks.concat([df1, df2], axis=1)
```

```
  letter  number  animal    name
0      a       1    bird   polly
1      b       2  monkey  george
```

Resolves #625, Closes #1009.

### CODE REVIEW

File: databricks/koalas/namespace.py [hunk 1]
```diff
@@ -1674,6 +1701,71 @@ def concat(objs, axis=0, join="outer", ignore_index=False):
...
+                for this_label, that_label in itertools.zip_longest(
+                    this_column_labels, that_column_labels
+                ):
+                    yield (kdf._kser_for(this_label), this_label)
+                    yield (kdf._kser_for(that_label), that_label)
+
+            for kser_or_kdf in objs[1:]:
+                if isinstance(kser_or_kdf, Series):
+                    # TODO: there is a corner case to optimize - when the series are from
+                    #   the same DataFrame.
```
Comment (itholic) [inline databricks/koalas/namespace.py]: too many whitespaces?

Comment (HyukjinKwon) [reply databricks/koalas/namespace.py]: It makes a todo as a sentence :-). 
Screen Shot 2020-03-18 at 7 02 05 PM
Screen Shot 2020-03-18 at 7 02 19 PM

File: databricks/koalas/tests/test_namespace.py [hunk 2]
```diff
@@ -128,3 +125,53 @@ def test_concat(self):
...
+                [pdf3[("X", "A")], pdf3[("X", "B")].rename("B")],
+            ),
+        ]
+
+        for ignore_index, join in itertools.product(ignore_indexes, joins):
+            for i, obj in enumerate(objs):
+                kdfs, pdfs = obj
+                with self.subTest(ignore_index=ignore_index, join=join, objs=obj):
+                    expected = ks.concat(kdfs, axis=1, ignore_index=ignore_index, join=join)
+                    actual = pd.concat(pdfs, axis=1, ignore_index=ignore_index, join=join)
```
Comment (ueshin) [inline databricks/koalas/tests/test_namespace.py]: nit: `actual = ks.concat(..)`, `expected = pd.concat(..)`?

File: databricks/koalas/tests/test_namespace.py [hunk 3]
```diff
@@ -128,3 +125,66 @@ def test_concat(self):
...
+        ]
+
+        for ignore_index, join in itertools.product(ignore_indexes, joins):
+            for obj in objs:
+                kdfs, pdfs = obj
+                with self.subTest(ignore_index=ignore_index, join=join, objs=obj):
+                    actual = ks.concat(kdfs, axis=1, ignore_index=ignore_index, join=join)
+                    expected = pd.concat(pdfs, axis=1, ignore_index=ignore_index, join=join)
+                    if ignore_index:
+                        expected.columns = list(map(str, actual.columns))
```
Comment (ueshin) [inline databricks/koalas/tests/test_namespace.py]: `expected.columns = list(map(str, expected.columns))`?

**Comment by HyukjinKwon** [comment]
In the last commits, I addressed some corner cases such as operations against single column index vs multi-column index, and fixed some nits.

File: databricks/koalas/tests/test_namespace.py [hunk 4]
```diff
@@ -128,3 +125,75 @@ def test_concat(self):
...
+            ([kdf3, kdf4], [pdf3, pdf4]),
+            ([kdf3[("X", "A")], kdf3[("X", "B")]], [pdf3[("X", "A")], pdf3[("X", "B")]],),
+            (
+                [kdf3[("X", "A")], kdf3[("X", "B")].rename("ABC")],
+                [pdf3[("X", "A")], pdf3[("X", "B")].rename("ABC")],
+            ),
+            (
+                [kdf3[("X", "A")].rename("ABC"), kdf3[("X", "B")]],
+                [pdf3[("X", "A")].rename("ABC"), pdf3[("X", "B")]],
+            ),
```
Comment (HyukjinKwon) [inline databricks/koalas/tests/test_namespace.py]: This two test cases were added to verify single index vs multi-index in the column.

File: databricks/koalas/tests/test_namespace.py [hunk 5]
```diff
@@ -128,3 +125,75 @@ def test_concat(self):
...
+            ValueError,
+            r"Only can inner \(intersect\) or outer \(union\) join the other axis.",
+            lambda: ks.concat([kdf.A, kdf4.B], join="", axis=1),
+        )
+
+        self.assertRaisesRegex(
+            ValueError,
+            r"Labels have to be unique; however, got duplicated labels \['A'\].",
+            lambda: ks.concat([kdf.A, kdf4.A], join="inner", axis=1),
+        )
```
Comment (HyukjinKwon) [inline databricks/koalas/tests/test_namespace.py]: This test also was added to verify duplicated column cases.

File: databricks/koalas/tests/test_namespace.py [hunk 6]
```diff
@@ -128,3 +125,82 @@ def test_concat(self):
...
+        kdf3.columns = columns
+
+        columns = pd.MultiIndex.from_tuples([("X", "C"), ("X", "D")])
+        pdf4.columns = columns
+        kdf4.columns = columns
+
+        pdf5 = pd.DataFrame({"A": [0, 2, 4], "B": [1, 3, 5]}, index=[1, 2, 3])
+        pdf6 = pd.DataFrame({"C": [1, 2, 3]}, index=[1, 3, 5])
+        kdf5 = ks.from_pandas(pdf5)
+        kdf6 = ks.from_pandas(pdf6)
```
Comment (HyukjinKwon) [inline databricks/koalas/tests/test_namespace.py]: Added another test case.

**Comment by HyukjinKwon** [comment]
Merged!
